framework/src/media/FileInputDataSource.cpp

```diff
@@ -96,46 +96,50 @@ bool FileInputDataSource::isPrepare()
 
 size_t FileInputDataSource::read(unsigned char *buf, size_t size)
 {
-	size_t ret;
+	size_t rlen = 0;
 	size_t readSize = size;
 	std::shared_ptr<Decoder> decoder = getDecoder();
 	if (decoder) {
-		size_t frames = 0;
-		/* TODO Currently it just return frames which decoded only ONCE if available */
-		ret = getDecodeFrames(buf, &frames);
-		medvdbg("decode frame : %d\n", frames);
-		if (ret > 0) {
-			return ret;
+		size_t frames = size;
+		rlen = getDecodeFrames(buf, &frames);
+		medvdbg("decode frame : %d/%d\n", rlen, size);
+		if (rlen == size) {
+			return rlen;
 		}
 
+		/* Set buf offset */
+		buf += rlen;
+		/* Calculate available space in 'buf' */
+		readSize -= rlen;
+
 		if (readSize > decoder->getAvailSpace()) {
 			readSize = decoder->getAvailSpace();
 		}
 	}
 
-	ret = fread(buf, sizeof(unsigned char), readSize, mFp);
 	/* If file position reaches end of file, return 0 */
 	if (feof(mFp)) {
 		medvdbg("eof!! stop!\n");
-		return 0;
+		return rlen;
 	}
 
+	size_t ret = fread(buf, sizeof(unsigned char), readSize, mFp);
```

Comment:
read -> feof check is correct I think.
In your case, fread failed due to eof, but it call pushData.

Reply:
`man fread`
RETUREN VALUE: `... If an error occurs, or the end of the file is reached, the return value is a short item count (or zero).`
	if (ret == 0) {
		return rlen;
	}
	if (decoder) {
		// push data ....
	}
So, if retrurn value is 0 (eof or error), we do not call pushData.
Actually, I don't understand why fread -> feof was used. 
Even if feof return 1 (true), fread may still have some data read (a short item count), which should be pushed.
Please help to tell the problem of my logic, thanks.

Reply:
Ok what you meaning is that below case right?
1. file size is 200byte, and read size is 8 byte
2. file pos reached 196byte
3. check eof
4. read 8 byte more
5. push back
6. check eof, return values

right?

Reply:
1. file size is 204 bytes, and read size is 8 bytes
2. after 25 times reading, we have read 200 bytes.
3. at the 26th time, fread return 4 bytes only, 
4. if check feof, it's true.
5. but we still need to push these 4 bytes data, isn't it?
so I move feof before fread.

Reply:
yeah you're right we need remained data.